crates/pink/pink-extension-runtime/src/local_cache.rs

```diff
@@ -0,0 +1,410 @@
+        use std::cell::RefCell;
+        thread_local! {
+            pub static GLOBAL_CACHE: RefCell<LocalCache> = RefCell::new(LocalCache::new());
+        }
+        GLOBAL_CACHE.with(move |cache| f(&mut cache.borrow_mut()))
+    } else {
+        use std::sync::Mutex;
+        pub static GLOBAL_CACHE: Mutex<LocalCache> = Mutex::new(LocalCache::new());
+        f(&mut GLOBAL_CACHE.lock().unwrap())
+    }
+}
+
+struct Storage {
+    // Sum of the size of all the keys and values.
+    size: usize,
+    max_size: usize,
+    kvs: BTreeMap<Vec<u8>, StorageValue>,
+}
+
+impl Storage {
+    fn new(max_size: usize) -> Self {
+        Self {
+            size: 0,
+            max_size,
+            kvs: Default::default(),
+        }
+    }
+
+    fn fit_size(&mut self) {
+        if self.size <= self.max_size {
+            return;
+        }
+        let map = std::mem::take(&mut self.kvs);
+
+        let mut kvs: Vec<_> = map
+            .into_iter()
+            .map(|(k, v)| (v.expire_at, (k, v)))
+            .collect();
+        kvs.sort_by_key(|(expire, _)| *expire);
+        self.kvs = kvs
+            .into_iter()
+            .filter_map(|(_, (k, v))| {
+                if self.size < self.max_size {
+                    return Some((k, v));
+                }
+                self.size -= k.len() + v.value.len();
+                None
+            })
+            .collect();
+    }
+
+    fn clear_expired(&mut self, now: u64) {
+        self.kvs.retain(|k, v| {
+            if v.expire_at > now {
+                true
+            } else {
+                self.size -= v.value.len() + k.len();
+                false
+            }
+        });
+    }
+
+    fn remove(&mut self, key: &[u8]) -> Option<Vec<u8>> {
+        let v = self.kvs.remove(key).map(|v| v.value);
+        if let Some(v) = &v {
+            self.size -= v.len() + key.len();
+        }
+        v
+    }
+
+    fn set(
+        &mut self,
+        key: Cow<[u8]>,
+        value: Cow<[u8]>,
+        lifetime: u64,
+    ) -> Result<(), StorageQuotaExceeded> {
+        _ = self.remove(key.as_ref());
+        let data_len = key.len() + value.len();
+        let mut store_size = self.size + data_len;
+        if store_size > self.max_size {
+            self.clear_expired(now());
+            store_size = self.size + data_len;
+            if store_size > self.max_size {
+                return Err(StorageQuotaExceeded);
+            }
+        }
+        self.size = store_size;
+        self.kvs.insert(
+            key.into_owned(),
+            StorageValue {
+                expire_at: now().saturating_add(lifetime),
+                value: value.into_owned(),
+            },
+        );
+        Ok(())
+    }
+
+    #[cfg(test)]
+    fn get(&self, key: &[u8]) -> Option<&StorageValue> {
+        self.kvs.get(key)
+    }
+}
+
+struct StorageValue {
+    // Expiration time in seconds since the first call to `now`.
+    expire_at: u64,
```

Comment:
Is it the unit timestamp or relative time?

Reply:
It is relative time, the seconds since the first calling of `fn now()`, where now is defined as:
```
fn now() -> u64 {
    static REF_TIME: Lazy<Instant> = Lazy::new(Instant::now);
    REF_TIME.elapsed().as_secs()
}
```

Reply:
So it's relative to the first time calling `now()` in the lifecycle of the pruntime process. Sounds a bit weird. What's the reason we don't use the absolute value?

Reply:
The Instant doesn't define an absolute value. It can only compare two Instants to get a Duration.
The reason for using Instant rather than SystemTime is that Instant is not effected by the OS time and never decreases.

Reply:
Got it. Make sense to me now